org.eclipse.xtext/src/org/eclipse/xtext/xtext/OverriddenValueInspector.java

```diff
@@ -160,9 +170,16 @@ public Boolean caseRuleCall(RuleCall object) {
 		if (!addVisited(parserRule))
 			return Boolean.FALSE;
 		Multimap<String, AbstractElement> prevAssignedFeatures = assignedFeatures;
-		assignedFeatures = newMultimap();
-		doSwitch(parserRule.getAlternatives());
-		for (String feature : assignedFeatures.keySet())
+		
+		// Cannot use #computeIfAbsent since we will call this recursively and that causes ConcurrentModificationExceptions
+		ImmutableMultimap<String, AbstractElement> assignedInCalledRule = assignedFeaturesAtEnd.get(parserRule);
+		if (assignedInCalledRule == null) {
+			assignedFeatures = newMultimap();
+			doSwitch(parserRule.getAlternatives());
+			assignedInCalledRule = ImmutableMultimap.copyOf(assignedFeatures);
+			assignedFeaturesAtEnd.put(parserRule, assignedInCalledRule);
```

Comment:
is this safe regarding entry creating and multithreading?

Reply:
It's instantiated during a single validation check-method (see XtextValidator). Therefore I'm sure that it'S thread-safe.